service/student_service_test.go

```diff
@@ -0,0 +1,363 @@
+package service_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"example.com/student-service/proto"
+	"example.com/student-service/repository/student"
+	"example.com/student-service/service"
+	"example.com/student-service/service/mocks"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/golang/mock/gomock"
+	_ "github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type fixedIDGenerator struct{}
+
+func (f fixedIDGenerator) GenerateID() string {
+	return "generated-id"
+}
+
+// StudentServiceTestSuite defines the test suite structure
+type StudentServiceTestSuite struct {
+	suite.Suite
+
+	ctrl     *gomock.Controller
+	mockRepo *mocks.MockRepository
+	mockTime *mocks.MockTimeProvider
+	mockTx   *mocks.MockTxManager
+	server   *service.StudentServer
+}
+
+func (s *StudentServiceTestSuite) SetupTest() {
+	s.ctrl = gomock.NewController(s.T())
+
+	s.mockRepo = mocks.NewMockRepository(s.ctrl)
+	s.mockTime = mocks.NewMockTimeProvider(s.ctrl)
+	s.mockTx = mocks.NewMockTxManager(s.ctrl)
+
+	idGen := fixedIDGenerator{}
+	s.server = service.NewStudentServer(s.mockRepo, s.mockTime, s.mockTx, idGen)
+}
+
+func (s *StudentServiceTestSuite) TestSuiteRuns() {
+	s.T().Log("Test suite is running")
+	s.True(true)
+}
+
+func (s *StudentServiceTestSuite) TearDownTest() {
+	s.ctrl.Finish()
+}
+
+func (s *StudentServiceTestSuite) TestCreateStudent_Success() {
+	ctx := context.Background()
+
+	req := &proto.CreateStudentRequest{
+		FirstName: "Ivan",
+		LastName:  "Petrov",
+		Grade:     9,
+	}
+	createdAt := time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC)
+	expectedStudent := student.Student{
+		ID:        "generated-id",
+		FirstName: "Ivan",
+		LastName:  "Petrov",
+		Grade:     9,
+		CreatedAt: createdAt,
+	}
+
+	s.mockTime.EXPECT().Now().Return(time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC))
+	s.mockTx.EXPECT().WithTransaction(gomock.Any(), gomock.Any()).DoAndReturn(
+		func(_ context.Context, fn func(context.Context) error) error {
+			return fn(ctx)
+		},
+	)
+	s.mockRepo.EXPECT().Create(ctx, gomock.Eq(expectedStudent)).Return("generated-id", nil)
+
+	resp, err := s.server.CreateStudent(ctx, req)
+
+	require.NoError(s.T(), err)
+	require.Equal(s.T(), "generated-id", resp.GetId())
+}
+
+func (s *StudentServiceTestSuite) TestCreateStudent_DBError() {
+	ctx := context.Background()
+	req := &proto.CreateStudentRequest{FirstName: "Fail", LastName: "Case", Grade: 10}
+
+	s.mockTime.EXPECT().Now().Return(time.Now()).AnyTimes()
+	s.mockTx.EXPECT().WithTransaction(gomock.Any(), gomock.Any()).DoAndReturn(func(_ context.Context, fn func(context.Context) error) error {
```

Comment:
lets make a helper s.expectTransaction() for it finally

Reply:
Also, fn(ctx) should use the context that you did not assigned to the variable -> _